gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/binary/types/MapSerializer.java

```diff
@@ -24,19 +24,50 @@
 import org.apache.tinkerpop.gremlin.structure.io.Buffer;
 
 import java.io.IOException;
+import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
 
 public class MapSerializer extends SimpleTypeSerializer<Map> {
+
     public MapSerializer() {
         super(DataType.MAP);
     }
 
+    @Override
+    public Map readValue(final Buffer buffer, final GraphBinaryReader context, final boolean nullable) throws IOException {
+        if (nullable) {
+            // save the reader index when we check if value is null or not
+            final int readerIdx = buffer.readerIndex();
+            final byte valueFlag = buffer.readByte();
+            if ((valueFlag & 1) == 1) {
+                return null;
+            }
+            if ((valueFlag & 2) == 2) {
+                return readOrderedMap(buffer, context);
+            }
+
+        }
+
+        return readValue(buffer, context);
+    }
+
     @Override
     protected Map readValue(final Buffer buffer, final GraphBinaryReader context) throws IOException {
         final int length = buffer.readInt();
 
+        final Map<Object,Object> result = new HashMap<>(length);
```

Comment:
Nit: Probably makes sense to swap this around as the previous behavior defaulted to LinkedHashMap.